src/iam.ts

```diff
@@ -1357,58 +1380,90 @@ export class IAM extends IAMBase {
     }
   }
 
+  private async approveRolePublishing({ subject, role, version }: { subject: string, role: string, version: number }) {
+    if (!this._signer) {
+      throw new Error(ERROR_MESSAGES.SIGNER_NOT_INITIALIZED);
+    }
+
+    const erc712_type_hash = id("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)");
+    const agreement_type_hash = id("Agreement(address subject,bytes32 role,uint256 version)");
+
+    const domainSeparator = keccak256(
+      defaultAbiCoder.encode(
+        ["bytes32", "bytes32", "bytes32", "uint256", "address"],
+        [erc712_type_hash, id("Claim Manager"), id("1.0"), (await this._provider.getNetwork()).chainId, this._claimManager.address]
+      )
+    );
+
+    const messageId = Buffer.from("1901", "hex");
+
+    const agreementHash = solidityKeccak256(
+      ["bytes", "bytes32", "bytes32"],
+      [
+        messageId,
+        domainSeparator,
+        keccak256(defaultAbiCoder.encode(
+          ["bytes32", "address", "bytes32", "uint256"],
+          [agreement_type_hash, subject, namehash(role), version]
+        ))
+      ]
+    );
+
+    return this._signer.signMessage(arrayify(
+      agreementHash
+    ));
+  }
+
   async createClaimRequest({
     issuer,
     claim,
-    subject
+    subject,
+    registrationTypes = [RegistrationTypes.OffChain]
   }: {
     issuer: string[];
-    claim: { claimType: string; fields: { key: string; value: string | number }[] };
+    claim: { claimType: string; claimTypeVersion: number; fields: { key: string; value: string | number }[] };
     subject?: string;
+    registrationTypes?: RegistrationTypes[];
   }) {
     if (!this._did) {
       throw new Error(ERROR_MESSAGES.USER_NOT_LOGGED_IN);
     }
     if (!subject) {
       subject = this._did;
     }
+    const { claimType: role, claimTypeVersion: version } = claim;
+    const token = await this.createPublicClaim({ data: claim, subject });
 
-    const [roleDefinition, didDocument] = await Promise.all([
-      this.getDefinition({
-        type: ENSNamespaceTypes.Roles,
-        namespace: claim.claimType
-      }),
-      this.getDidDocument({ did: subject, includeClaims: true })
-    ]);
-
-    if (!roleDefinition) {
-      throw new Error(ERROR_MESSAGES.ROLE_NOT_EXISTS);
+    if (registrationTypes.includes(RegistrationTypes.OffChain)) {
```

Comment:
This is probably okay for now, as the `ClaimManager` contract will eventually do it, but maybe we should verify their on-chain preconditions prior to them requesting as well.